source/simulator/nullspace.cc

```diff
@@ -81,135 +90,184 @@ namespace aspect
-               ExcNotImplemented());
-
-        net_rotations_translations.resize(funcs.size());
-        for (unsigned int i=0; i<funcs.size(); ++i)
-          {
-            // for each of the null space dimensions, set up
-            // a vector, fill it with an element of the null space,
-            // and normalize it
-            net_rotations_translations[i].reinit(
-              introspection.index_sets.system_partitioning[introspection.block_indices.velocities],
-              mpi_communicator);
-            interpolate_onto_velocity_system(*funcs[i],
-                                             net_rotations_translations[i]);
-            net_rotations_translations[i] /= net_rotations_translations[i].l2_norm();
-          }
+    QGauss<dim> quadrature(parameters.stokes_velocity_degree+1);
+    FEValues<dim> fe(mapping, finite_element, quadrature,
+                     UpdateFlags(update_quadrature_points | update_JxW_values | update_values));
 
-      }
-  }
+    typename DoFHandler<dim>::active_cell_iterator cell;
+    std::vector<Point<dim> > q_points(quadrature.size());
+    std::vector<Vector<double> > fe_vals(quadrature.size(), Vector<double>(finite_element.n_components()));
 
+    Tensor<1,dim> local_momentum;
+    double local_mass;
 
+    //loop over all local cells
+    for (cell = dof_handler.begin_active(); cell != dof_handler.end(); ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe.reinit (cell);
+          q_points = fe.get_quadrature_points();
+          fe.get_function_values(relevant_dst, fe_vals);
 
-  template <int dim>
-  void Simulator<dim>::remove_nullspace(LinearAlgebra::BlockVector &relevant_dst,
-                                        LinearAlgebra::BlockVector &tmp_distributed_stokes)
-  {
-    if (parameters.nullspace_removal & NullspaceRemoval::net_rotation ||
-        parameters.nullspace_removal & NullspaceRemoval::net_translation )
+          // get the density at each quadrature point
+          typename MaterialModel::Interface<dim>::MaterialModelInputs in(q_points.size(),
+                                                                parameters.n_compositional_fields);
+          typename MaterialModel::Interface<dim>::MaterialModelOutputs out(q_points.size(),
+                                                                  parameters.n_compositional_fields);
+          for (unsigned int i=0; i< q_points.size(); i++)
+            {
+              in.pressure[i] = fe_vals[i][introspection.component_indices.pressure];
+              in.temperature[i] = fe_vals[i][introspection.component_indices.temperature];
+              for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+                in.composition[i][c] = fe_vals[i][introspection.component_indices.compositional_fields[c]];
+              in.position[i] = q_points[i];
+
+            }
+          material_model->evaluate(in, out);
```

Comment:
That's a potentially expensive call, but you won't need it if `use_constant_density==true`.